test/core/client_channel/resolvers/custom_resolver_test.cc

```diff
@@ -0,0 +1,359 @@
+    }
+    gpr_free(balancer_name);
+    gpr_free(uri_string);
+  }
+  return addresses;
+}
+
+static on_resolution_arg create_on_resolution_arg(const Addresses& results) {
+  on_resolution_arg on_res_arg;
+  memset(&on_res_arg, 0, sizeof(on_res_arg));
+  on_res_arg.expected_resolver_result =
+      grpc_addresses_copy_lb_addresses_for_test(results.get());
+  gpr_event_init(&on_res_arg.ev);
+  return on_res_arg;
+}
+
+class Resolver {
+ public:
+  explicit Resolver(grpc_resolver_observer* observer) : observer_(observer) {}
+  virtual ~Resolver() { grpc_resolver_observer_destroy(observer_); }
+
+  virtual void RequestReresolution() = 0;
+
+  static void RequestReresolutionWrapper(void* user_data) {
+    static_cast<Resolver*>(user_data)->RequestReresolution();
+  }
+  static void DestroyWrapper(void* user_data) {
+    delete static_cast<Resolver*>(user_data);
+  }
+
+ protected:
+  void SetAddresses(const Addresses& addresses) {
+    grpc_resolver_observer_set_addresses(observer_, addresses.get());
+  }
+
+ private:
+  grpc_resolver_observer* observer_;
+  grpc_addresses* reresolution_response_ = nullptr;
+};
+
+class ResolverFactory {
+ public:
+  ResolverFactory() {}
+  virtual ~ResolverFactory() {}
+
+  virtual Resolver* Resolve(grpc_resolver_args* args,
+                            grpc_resolver_observer* observer) = 0;
+
+  static void* ResolveWrapper(void* user_data, grpc_resolver_args* args,
+                              grpc_resolver_observer* observer) {
+    return static_cast<ResolverFactory*>(user_data)->Resolve(args, observer);
+  }
+  static void DestroyWrapper(void* user_data) {
+    delete static_cast<ResolverFactory*>(user_data);
+  }
+};
+
+class FakeResolver : public Resolver {
+ public:
+  FakeResolver(grpc_resolver_args* args, grpc_resolver_observer* observer)
+      : Resolver(observer) {}
+
+  void RequestReresolution() {
+    if (reresolution_response_ != nullptr) {
+      SetAddresses(reresolution_response_);
+    }
+  }
+
+  void SetResponse(const Addresses& addresses) { SetAddresses(addresses); }
+
+  void SetReresolutionResponse(Addresses addresses) {
+    reresolution_response_ = std::move(addresses);
+  }
+
+ private:
+  Addresses reresolution_response_ = nullptr;
+};
+
+class FakeResolverFactory : public ResolverFactory {
+ public:
+  FakeResolverFactory(FakeResolver** resolver) : resolver_(resolver) {}
+
+  Resolver* Resolve(grpc_resolver_args* args,
+                    grpc_resolver_observer* observer) {
+    GPR_ASSERT(resolver_ != nullptr);
+    auto resolver = new FakeResolver(args, observer);
+    *resolver_ = resolver;
+    resolver_ = nullptr;
+    return resolver;
+  }
+
+ private:
+  FakeResolver** resolver_;
+};
+
+static grpc_core::OrphanablePtr<grpc_core::Resolver> build_custom_resolver(
+    const char* scheme, grpc_combiner* combiner) {
+  grpc_core::ResolverFactory* factory =
+      grpc_core::ResolverRegistry::LookupResolverFactory(scheme);
+  grpc_channel_args channel_args = {0, nullptr};
```

Comment:
If there are no args, then there's no need for this; just set `args.args` to nullptr.